synapseclient/models/services/storable_entity_components.py

```diff
@@ -51,59 +79,86 @@ async def store_entity_components(
                 )
             )
 
-    if (
-        hasattr(root_resource, "annotations")
-        and root_resource.annotations is not None
-        and (
-            last_persistent_instance is None
-            or last_persistent_instance.annotations != root_resource.annotations
-        )
-    ):
-        tasks.append(
-            asyncio.create_task(
-                Annotations(
-                    id=root_resource.id,
-                    etag=root_resource.etag,
-                    annotations=root_resource.annotations,
-                ).store(synapse_client=synapse_client)
+    tasks.append(
+        asyncio.create_task(
+            _store_activity_and_annotations(
+                root_resource, synapse_client=synapse_client
             )
         )
+    )
 
     try:
+        tasks = [wrap_coroutine(task) for task in tasks]
         for task in asyncio.as_completed(tasks):
             result = await task
-            if isinstance(result, Annotations):
-                root_resource.annotations = result.annotations
-                root_resource.etag = result.etag
-            elif result.__class__.__name__ == "Folder":
-                pass
-            elif result.__class__.__name__ == "File":
-                pass
-            else:
-                if isinstance(result, BaseException):
-                    Synapse.get_client(synapse_client=synapse_client).logger.exception(
-                        result
-                    )
-                    raise result
-                raise ValueError(f"Unknown type: {type(result)}", result)
+            _resolve_store_task(
+                result=result,
+                failure_strategy=failure_strategy,
+                synapse_client=synapse_client,
+            )
     except Exception as ex:
         Synapse.get_client(synapse_client=synapse_client).logger.exception(ex)
-        raise ex
+        if failure_strategy == FailureStrategy.RAISE_EXCEPTION:
+            raise ex
 
-    re_read_required = await _store_activity(
-        root_resource, synapse_client=synapse_client
-    )
+    return re_read_required
+
+
+def _resolve_store_task(
+    result: Union[bool, "Folder", "File", BaseException],
```

Comment:
Are the "Folder" / "File" optional values correct ? This `result` arg later gets called to get the `.__class__.__name__` attribute which would just be `'str'` for these. Unless there is something I'm missing about type hints here

Reply:
It is correct here - The problem that this is solving is an issue of circular dependencies. `isinstance` only works if you are passing an initialized reference to the class - But I was unable to import it at the top of the file (circular dependency) or under `if TYPE_CHECKING` (Not initialized).

I could have done a lazy import (ie: Importing in the method) - But this also works.